pkg/ipcache/cidr.go

```diff
@@ -207,13 +215,16 @@ func (ipc *IPCache) releaseCIDRIdentities(ctx context.Context, prefixes []netip.
 			}).WithError(err).Warning("Unable to release CIDR identity. Ignoring error. Identity may be leaked")
 		}
 		if released {
+			deletedIDs[id.ID] = id.LabelArray
 			toDelete = append(toDelete, prefix)
 		}
 	}
 
 	for _, prefix := range toDelete {
 		ipc.deleteLocked(prefix.String(), source.Generated)
 	}
+	// Remove any deleted identities from the policy engine.
+	ipc.UpdatePolicyMaps(ctx, nil, deletedIDs)
```

Comment:
Note that this happens while holding the ipcache lock, specifically to mitigate a race condition per the comment at the top of this function (see also commit logs for the corresponding lines, https://github.com/cilium/cilium/commit/40e13ea2a5a944a45761fc433c4c971536957f4b). This means it's possible for this goroutine to be holding the ipcache lock and grab the endpoint lock, then the selectorcache lock.

I believe that you recently changed the locking patterns from the endpoint side to ensure that the endpoint never holds its own lock while attempting to push policy-triggered ipcache updates through to the ipcache. So this _probably_ means that there's no deadlock here. Worth double-checking from your side though.

Reply:
It's even more academic, since regenerating an endpoint / policy calculation can't trigger legacy synchronous ipcache calls anymore. That would have been as a result of FQDN policy, but those use the asynch APIs now. This method is only called by the k8s service handler, which doesn't hold any locks.